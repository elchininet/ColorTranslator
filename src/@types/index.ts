export type NumberOrString = number | string;

export interface HEXObject {
    r: string;
    g: string;
    b: string;
    a?: string;
}

export interface RGBObject {
    r: number;
    g: number;
    b: number;
    a?: number;
}

export interface RGBObjectGeneric {
    r: NumberOrString;
    g: NumberOrString;
    b: NumberOrString;
    a?: NumberOrString;
}

export interface HSLObject {
    h: number;
    s: number;
    l: number;
    a?: number;
}

export interface HSLObjectGeneric {
    h: number;
    s: NumberOrString;
    l: NumberOrString;
    a?: number;
}

export interface CMYKObject {
    c: number;
    m: number;
    y: number;
    k: number;
    a?: number;
}

export interface CMYKObjectGeneric {
    c: NumberOrString;
    m: NumberOrString;
    y: NumberOrString;
    k: NumberOrString;
    a?: number;
}

export interface RYBObject {
    r: number;
    y: number;
    b: number;
    a?: number;
}

export type RGYBObject = RGBObject | RYBObject;

export type Color = RGBObjectGeneric | HSLObjectGeneric | CMYKObjectGeneric;
export type ColorWithoutCMYK = RGBObjectGeneric | HSLObjectGeneric;

export type ColorInput = string | Color;
export type ColorInputWithoutCMYK = string | ColorWithoutCMYK;
export type HEXOutput = string | HEXObject;
export type RGBOutput = string | RGBObject;
export type HSLOutput = string | HSLObject;
export type CMYKOutput = string | CMYKObject;
export type ColorOutput = HEXOutput | RGBOutput | HSLOutput;

export interface ObjectProps<T> {
    [key: string]: T;
}

export interface Options {
    decimals: number;
    legacyCSS: boolean;
    spacesAfterCommas: boolean;
}

export type InputOptions = Partial<Options>;