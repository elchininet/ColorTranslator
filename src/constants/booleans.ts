export const DEFAULT_COLOR_LEVEL_4 = false;