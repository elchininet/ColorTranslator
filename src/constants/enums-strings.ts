import { Harmony, Mix } from './enums';

export type HarmonyString = `${Harmony}`;
export type MixString = `${Mix}`;