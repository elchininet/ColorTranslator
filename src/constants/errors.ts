export const ERRORS = {
    NOT_ACCEPTED_STRING_INPUT: 'The provided string color doesn\'t have a correct format',
    NOT_ACCEPTED_OBJECT_INPUT: 'The provided color object doesn\'t have the proper keys or format',
    NOT_A_VALID_RELATIVE_COLOR: 'The provided relative color is not valid'
};