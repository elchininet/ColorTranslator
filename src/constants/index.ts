export * from './numbers';
export * from './enums';
export * from './regexps';
export * from './errors';
