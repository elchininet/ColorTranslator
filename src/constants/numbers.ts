export const MAX_DECIMALS = 6;
export const MIN_DECIMALS = 2;
