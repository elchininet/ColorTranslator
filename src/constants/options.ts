import {
    Options,
    AnglesUnitEnum,
    ColorUnitEnum,
    CMYKFunctionEnum
} from '@types';
import { MAX_DECIMALS } from './numbers';

export const DEFAULT_OPTIONS: Options = {
    decimals: MAX_DECIMALS,
    legacyCSS: false,
    spacesAfterCommas: false,
    anglesUnit: AnglesUnitEnum.NONE,
    rgbUnit: ColorUnitEnum.NONE,
    labUnit: ColorUnitEnum.NONE,
    cmykUnit: ColorUnitEnum.PERCENT,
    alphaUnit: ColorUnitEnum.NONE,
    cmykFunction: CMYKFunctionEnum.DEVICE_CMYK
};