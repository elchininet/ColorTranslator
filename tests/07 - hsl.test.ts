import { ColorTranslator } from '../src';
import { COLORS } from './tests.constants';

COLORS.forEach((item): void => {

    describe(`HSL tests for ${item.HSL}`, (): void => {

        const HSL = { ...item.HSLObject };
        const HSLA = { ...item.HSLObject };
        const HSLPercentages = { ...HSL, S: HSL.S + '%', L: HSL.L + '%' };
        const HSLAPercentages = { ...HSLA, S: HSLA.S + '%', L: HSLA.L + '%' };
        const HSLBigHue = { ...HSL, H: HSL.H + 360 };
        const HSLABigHue = { ...HSL, H: HSL.H + 360 };
        const HSLLowHue = { ...HSL, H: HSL.H - 360 };
        const HSLALowHue = { ...HSL, H: HSL.H - 360 };
        const options = { decimals: 0 };
        const legacyOptions = { ...options, legacyCSS: true };

        it(`HSL with percentages ${JSON.stringify(HSLPercentages)}`, (): void => {

            expect(ColorTranslator.toRGB(HSLPercentages, options)).toBe(item.RGB);
            expect(ColorTranslator.toRGB(HSLPercentages, legacyOptions)).toBe(item.RGBLegacy);
            expect(ColorTranslator.toRGBObject(HSLPercentages, options)).toMatchObject(item.RGBObject);

            expect(ColorTranslator.toHSL(HSLPercentages, options)).toBe(item.HSL);
            expect(ColorTranslator.toHSL(HSLPercentages, legacyOptions)).toBe(item.HSLLegacy);
            expect(ColorTranslator.toHSLObject(HSLPercentages, options)).toMatchObject(item.HSLObject);

            expect(ColorTranslator.toRGB(HSLPercentages)).toMatchSnapshot();
            expect(ColorTranslator.toHSL(HSLPercentages)).toMatchSnapshot();

        });

        it(`HSLA with percentages ${JSON.stringify(HSLAPercentages)}`, (): void => {

            expect(ColorTranslator.toRGBA(HSLAPercentages, options)).toBe(item.RGBA);
            expect(ColorTranslator.toRGBA(HSLAPercentages, legacyOptions)).toBe(item.RGBALegacy);
            expect(ColorTranslator.toRGBAObject(HSLAPercentages, options)).toMatchObject(item.RGBAObject);

            expect(ColorTranslator.toHSLA(HSLAPercentages, options)).toBe(item.HSLA);
            expect(ColorTranslator.toHSLA(HSLAPercentages, legacyOptions)).toBe(item.HSLALegacy);
            expect(ColorTranslator.toHSLAObject(HSLAPercentages, options)).toMatchObject(item.HSLAObject);

            expect(ColorTranslator.toRGBA(HSLAPercentages)).toMatchSnapshot();
            expect(ColorTranslator.toHSLA(HSLAPercentages)).toMatchSnapshot();

        });

        it(`HSL big hue ${JSON.stringify(HSLBigHue)}`, (): void => {

            expect(ColorTranslator.toRGB(HSLBigHue, options)).toBe(item.RGB);
            expect(ColorTranslator.toRGB(HSLBigHue, legacyOptions)).toBe(item.RGBLegacy);
            expect(ColorTranslator.toRGBObject(HSLBigHue, options)).toMatchObject(item.RGBObject);

            expect(ColorTranslator.toHSL(HSLBigHue, options)).toBe(item.HSL);
            expect(ColorTranslator.toHSL(HSLBigHue, legacyOptions)).toBe(item.HSLLegacy);
            expect(ColorTranslator.toHSLObject(HSLBigHue, options)).toMatchObject(item.HSLObject);

        });

        it(`HSLA big hue ${JSON.stringify(HSLABigHue)}`, (): void => {

            expect(ColorTranslator.toRGBA(HSLABigHue, options)).toBe(item.RGBA);
            expect(ColorTranslator.toRGBA(HSLABigHue, legacyOptions)).toBe(item.RGBALegacy);
            expect(ColorTranslator.toRGBAObject(HSLABigHue, options)).toMatchObject(item.RGBAObject);

            expect(ColorTranslator.toHSLA(HSLABigHue, options)).toBe(item.HSLA);
            expect(ColorTranslator.toHSLA(HSLABigHue, legacyOptions)).toBe(item.HSLALegacy);
            expect(ColorTranslator.toHSLAObject(HSLABigHue, options)).toMatchObject(item.HSLAObject);

        });

        it(`HSL low hue ${JSON.stringify(HSLLowHue)}`, (): void => {

            expect(ColorTranslator.toRGB(HSLLowHue, options)).toBe(item.RGB);
            expect(ColorTranslator.toRGB(HSLLowHue, legacyOptions)).toBe(item.RGBLegacy);
            expect(ColorTranslator.toRGBObject(HSLLowHue, options)).toMatchObject(item.RGBObject);

            expect(ColorTranslator.toHSL(HSLLowHue, options)).toBe(item.HSL);
            expect(ColorTranslator.toHSL(HSLLowHue, legacyOptions)).toBe(item.HSLLegacy);
            expect(ColorTranslator.toHSLObject(HSLLowHue, options)).toMatchObject(item.HSLObject);

        });

        it(`HSLA low hue ${JSON.stringify(HSLALowHue)}`, (): void => {

            expect(ColorTranslator.toRGBA(HSLALowHue, options)).toBe(item.RGBA);
            expect(ColorTranslator.toRGBA(HSLALowHue, legacyOptions)).toBe(item.RGBALegacy);
            expect(ColorTranslator.toRGBAObject(HSLALowHue, options)).toMatchObject(item.RGBAObject);

            expect(ColorTranslator.toHSLA(HSLALowHue, options)).toBe(item.HSLA);
            expect(ColorTranslator.toHSLA(HSLALowHue, legacyOptions)).toBe(item.HSLALegacy);
            expect(ColorTranslator.toHSLAObject(HSLALowHue, options)).toMatchObject(item.HSLAObject);

        });

    });

});