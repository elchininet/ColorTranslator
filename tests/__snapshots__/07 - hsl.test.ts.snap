// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HSL tests for hsl(0 0% 0%) HSL with percentages {"H":0,"S":"0%","L":"0%"} 1`] = `"rgb(0 0 0)"`;

exports[`HSL tests for hsl(0 0% 0%) HSL with percentages {"H":0,"S":"0%","L":"0%"} 2`] = `"hsl(0 0% 0%)"`;

exports[`HSL tests for hsl(0 0% 0%) HSLA with percentages {"H":0,"S":"0%","L":"0%"} 1`] = `"rgb(0 0 0 / 1)"`;

exports[`HSL tests for hsl(0 0% 0%) HSLA with percentages {"H":0,"S":"0%","L":"0%"} 2`] = `"hsl(0 0% 0% / 1)"`;

exports[`HSL tests for hsl(0 0% 50%) HSL with percentages {"H":0,"S":"0%","L":"50%"} 1`] = `"rgb(127.5 127.5 127.5)"`;

exports[`HSL tests for hsl(0 0% 50%) HSL with percentages {"H":0,"S":"0%","L":"50%"} 2`] = `"hsl(0 0% 50%)"`;

exports[`HSL tests for hsl(0 0% 50%) HSLA with percentages {"H":0,"S":"0%","L":"50%"} 1`] = `"rgb(127.5 127.5 127.5 / 1)"`;

exports[`HSL tests for hsl(0 0% 50%) HSLA with percentages {"H":0,"S":"0%","L":"50%"} 2`] = `"hsl(0 0% 50% / 1)"`;

exports[`HSL tests for hsl(0 0% 100%) HSL with percentages {"H":0,"S":"0%","L":"100%"} 1`] = `"rgb(255 255 255)"`;

exports[`HSL tests for hsl(0 0% 100%) HSL with percentages {"H":0,"S":"0%","L":"100%"} 2`] = `"hsl(0 0% 100%)"`;

exports[`HSL tests for hsl(0 0% 100%) HSLA with percentages {"H":0,"S":"0%","L":"100%"} 1`] = `"rgb(255 255 255 / 1)"`;

exports[`HSL tests for hsl(0 0% 100%) HSLA with percentages {"H":0,"S":"0%","L":"100%"} 2`] = `"hsl(0 0% 100% / 1)"`;

exports[`HSL tests for hsl(0 100% 50%) HSL with percentages {"H":0,"S":"100%","L":"50%"} 1`] = `"rgb(255 0 0)"`;

exports[`HSL tests for hsl(0 100% 50%) HSL with percentages {"H":0,"S":"100%","L":"50%"} 2`] = `"hsl(0 100% 50%)"`;

exports[`HSL tests for hsl(0 100% 50%) HSLA with percentages {"H":0,"S":"100%","L":"50%"} 1`] = `"rgb(255 0 0 / 1)"`;

exports[`HSL tests for hsl(0 100% 50%) HSLA with percentages {"H":0,"S":"100%","L":"50%"} 2`] = `"hsl(0 100% 50% / 1)"`;

exports[`HSL tests for hsl(120 100% 50%) HSL with percentages {"H":120,"S":"100%","L":"50%"} 1`] = `"rgb(0 255 0)"`;

exports[`HSL tests for hsl(120 100% 50%) HSL with percentages {"H":120,"S":"100%","L":"50%"} 2`] = `"hsl(120 100% 50%)"`;

exports[`HSL tests for hsl(120 100% 50%) HSLA with percentages {"H":120,"S":"100%","L":"50%"} 1`] = `"rgb(0 255 0 / 1)"`;

exports[`HSL tests for hsl(120 100% 50%) HSLA with percentages {"H":120,"S":"100%","L":"50%"} 2`] = `"hsl(120 100% 50% / 1)"`;

exports[`HSL tests for hsl(240 100% 50%) HSL with percentages {"H":240,"S":"100%","L":"50%"} 1`] = `"rgb(0 0 255)"`;

exports[`HSL tests for hsl(240 100% 50%) HSL with percentages {"H":240,"S":"100%","L":"50%"} 2`] = `"hsl(240 100% 50%)"`;

exports[`HSL tests for hsl(240 100% 50%) HSLA with percentages {"H":240,"S":"100%","L":"50%"} 1`] = `"rgb(0 0 255 / 1)"`;

exports[`HSL tests for hsl(240 100% 50%) HSLA with percentages {"H":240,"S":"100%","L":"50%"} 2`] = `"hsl(240 100% 50% / 1)"`;

exports[`HSL tests for hsl(300 100% 50%) HSL with percentages {"H":300,"S":"100%","L":"50%"} 1`] = `"rgb(255 0 255)"`;

exports[`HSL tests for hsl(300 100% 50%) HSL with percentages {"H":300,"S":"100%","L":"50%"} 2`] = `"hsl(300 100% 50%)"`;

exports[`HSL tests for hsl(300 100% 50%) HSLA with percentages {"H":300,"S":"100%","L":"50%"} 1`] = `"rgb(255 0 255 / 1)"`;

exports[`HSL tests for hsl(300 100% 50%) HSLA with percentages {"H":300,"S":"100%","L":"50%"} 2`] = `"hsl(300 100% 50% / 1)"`;
